docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,213 @@
+   client, which exposes application methods to the node as direct function
+   calls.
+
+2. **Socket protocol**: Out-of-process applications may export the ABCI service
+   via a custom socket protocol that sends requests and responses over a
+   Unix-domain or TCP socket connection as length-prefixed protocol buffers.
+   In Tendermint, this is handled by the [socket client][socket-client].
+
+3. **gRPC**: Out-of-process applications may export the ABCI service via gRPC.
+   In Tendermint, this is handled by the [gRPC client][grpc-client].
+
+Both the out-of-process options (2) and (3) have a long history in Tendermint.
+The beginnings of the gRPC client were added in [May 2016][abci-start] when
+ABCI was still hosted in a separate repository, and the socket client (formerly
+called the "remote client") was part of ABCI from its inception in November
+2015.
+
+At that time when ABCI was first being developed, the gRPC project was very new
+(it launched Q4 2015) and it was not an obvious choice for use in Tendermint.
+It took a while before the language coverage and quality of gRPC reached a
+point where it could be a viable solution for out-of-process applications.  For
+that reason, it made sense for the initial design of ABCI to focus on a custom
+protocol for out-of-process applications.
+
+## Problem Statement
+
+For practical reasons, ABCI needs an interprocess communication option to
+support applications not written in Go. The two practical options are RPC and
+FFI, and for operational reasons an RPC mechanism makes more sense.
+
+The socket protocol has not changed all that substantially since its original
+design, and has the advantage of being simple to implement in almost any
+reasonable language.  However, its simplicity includes some limitations that
+have had a negative impact on the stability and performance of out-of-process
+applications using it. In particular:
+
+- The protocol lacks request identifiers, so the client and server must
+  complete requests in strict FIFO order. Even if the client issues requests
+  that have no dependency on each other, the protocol has no way except order
+  of issue to map responses to requests. This strictly limits the application's
+  ability to process requests concurrently, and has been a source of complaints
+  from some network operators on that basis.
+
+- The protocol lacks method identifiers, so the only way for the client and
+  server to understand which operation is requested is to dispatch on the type
+  of the request and response payloads. For responses, this means that [any
+  error condition is terminal not only to the request, but to the entire ABCI
+  client](https://github.com/tendermint/tendermint/blob/master/abci/client/socket_client.go#L149).
+
+  The historical intent of terminating for any error seems to have been that
+  all ABCI errors are unrecoverable and hence protocol fatal.  In practice,
```

Comment:
Yes, that does seem to have been the motivation. I'll add a few more words to that effect.

That decision was motivated by a sensible principle, but wound up having unfortunate engineering consequences: Treating all errors as protocol fatal conflated two different concerns of error reporting, viz., (1) semantic (something wrong in the content of the protocol, an error for which divergence is a risk) and (2) mechanical (e.g., timeouts, resource exhaustion, lost connection, descriptor saturation, full pipes, etc.).

In almost all cases, mechanical errors (2) should at least be "recovered" to a well-defined point in the program where they can be handled consistently.  Even if the handler does the same thing (e.g., exit the node rather than risk letting the application continue with a corrupt state) the value of handling that explicitly is important for consistency. 

When, as now, the application simply crashes for an error, we see different recovery behaviours even from the same application depending on whether it is running in-process (panic also panics the node) or out-of-process (panic may or may not be detected at all till later when a connection write fails). Shared process fate is tricky to get right—and we didn't in this case.

Not separating these concerns is one of the main reasons we now have unfortunate concurrency limitations even for _correct_ applications.

Reply:
I added a separate note, this was too much to inline. 🙂